### ISSUE TITLE
Achieve parity with the Python optimizer

### ISSUE THREAD

**Issue by dae** [open-spaced-repetition/fsrs-rs#88]
Currently the generated weights from this crate are slightly behind the Python optimizer. Not urgent, but in the long run, it would be nice if this crate could perform as well.

https://github.com/open-spaced-repetition/fsrs-benchmark#weighted-by-number-of-reviews

Any thoughts on what might be driving the differences?

**Comment by L-M-Sherlock** [comment]
1. The current outlier filter only remove those outliers in pretrain-set. Py version remove the them in the train-set, too: https://github.com/open-spaced-repetition/fsrs-optimizer/blob/94be1a22d93e84e121186ea7181f55bc826639d3/src/fsrs_optimizer/fsrs_optimizer.py#L654-L664
2. The Py version split the trainset and optimizer the weights for each split and calculate the average weights as the final result: thttps://github.com/open-spaced-repetition/fsrs-optimizer/blob/94be1a22d93e84e121186ea7181f55bc826639d3/src/fsrs_optimizer/fsrs_optimizer.py#L1050-L1065
3. The Py version applied [additive smoothing](https://github.com/open-spaced-repetition/fsrs4anki/pull/358) to the pretrain: https://github.com/open-spaced-repetition/fsrs-optimizer/blob/94be1a22d93e84e121186ea7181f55bc826639d3/src/fsrs_optimizer/fsrs_optimizer.py#L847-L849

**Comment by user1823** [comment]
> The current outlier filter only remove those outliers in pretrain-set. The py version removes the them in the train-set, too

In my opinion, the current outlier filter is too aggressive (it removes too many reviews). This is fine for the pretrain function (because the pretrain method is fragile). However, the train function should have access to more reviews. So, I think that we should build a less aggressive outlier filter for the train function.

There is no hurry in doing this. But, I wanted to make this point here so that you don't waste time and effort in applying the current outlier filter to the trainset.

**Comment by user1823** [comment]
The python optimizer selects the parameters from the epoch with the minimum loss. I don't think that the rust optimizer behaves similarly.

Related commit: https://github.com/open-spaced-repetition/fsrs4anki/commit/1719daee92bd33bfbc65fe986ae0483e6299b074

**Comment by L-M-Sherlock** [comment]
> The python optimizer selects the parameters from the epoch with the minimum loss. I don't think that the rust optimizer behaves similarly.

I also find out the difference, but the framework also doesn't also me to implement a similar feature.

**Comment by dae** [comment]
We could always switch to our own training loop if need be. @nathanielsimard I presume something like this is not supported out of the box at the moment?

**Comment by nathanielsimard** [comment]
I'm going to prioritize adding early stoping within `burn-train`, I think it's time we support that feature!

**Comment by nathanielsimard** [comment]
Burn Issue: https://github.com/burn-rs/burn/issues/841

**Comment by user1823** [comment]
In case you aren't already aware, I wanted to let you know that Burn now supports early stopping when training, implemented in https://github.com/burn-rs/burn/pull/878. Probably, now is the time to implement early stopping in the Rust version of the FSRS optimizer.

By the way, the comparison between the Python and the Rust versions of the optimizer is somewhat unfair.

The Python version filters several reviews before (training and) evaluation which are not filtered out by the Rust version. Examples include:
- Reviews of cards that are filtered by the outlier filter in pre-train.
- Reviews of cards with incomplete revlogs.

In my opinion, evaluation should happen for these reviews, which means that the behavior of the Rust version is desirable. However, we should keep in mind that it is inconsistent with the Python version.

**Comment by dae** [comment]
https://github.com/burn-rs/burn/pull/882 will need to be merged before we can update burn

**Comment by L-M-Sherlock** [comment]
> The Python version filters several reviews before (training and) evaluation which are not filtered out by the Rust version. Examples include:

The python version doesn't filter any reviews before training and evaluation in the benchmark. Here is the code:

https://github.com/open-spaced-repetition/fsrs-benchmark/blob/8aacff613116d70ab21325628235a38024d57d48/script.py#L120-L148
